File: .github/probes/macos-filesystem.py
"""Finite synthetic native filesystem tests; experiment branch only."""
import json
import os
import pathlib
import plistlib
import shutil
import stat
import subprocess
import sys
import tempfile

assert os.geteuid() == 0 and sys.platform == 'darwin'
assert os.environ.get('GITHUB_ACTIONS') == 'true'
base = pathlib.Path(tempfile.mkdtemp(prefix='crucible-fs-probe-')).resolve(strict=True)
os.chmod(base, 0o755)
worker = base / 'worker'
shutil.copyfile(pathlib.Path(sys.argv[1]).resolve(strict=True), worker)
os.chmod(worker, 0o755)
image = base / 'fixture.dmg'
mount = base / 'volume'
mount.mkdir()
print('FIXTURE ' + str(base), flush=True)


def command(argv, allowed=(0,), timeout=30):
    print('COMMAND ' + json.dumps([str(x) for x in argv]), flush=True)
    with tempfile.TemporaryFile() as output:
        proc = subprocess.Popen(argv, cwd=base, stdin=subprocess.DEVNULL,
                                stdout=output, stderr=subprocess.STDOUT,
                                close_fds=True, start_new_session=True,
                                env={'PATH': '/usr/bin:/bin:/usr/sbin:/sbin', 'LANG': 'C'})
        try:
            result = proc.wait(timeout=timeout)
        except subprocess.TimeoutExpired:
            proc.kill()
            try:
                proc.wait(timeout=3)
                print('TIMEOUT direct_process_reaped=1 fixture_quarantined=1', flush=True)
            except subprocess.TimeoutExpired:
                print('TIMEOUT direct_process_reaped=0 fixture_quarantined=1', flush=True)
            raise RuntimeError('No further fixture operations after uncertainty')
        size = output.tell()
        output.seek(0)
        data = output.read(32769)
        assert size <= 32768 and len(data) <= 32768, 'output bound'
    print(data.decode(errors='replace'), end='', flush=True)
    assert result in allowed, (str(argv[0]), result)
    return result, data


def owned_file(path, data=b'fixture'):
    path.write_bytes(data)
    os.chown(path, 60000, 60000)
    os.chmod(path, 0o666)


def make_case(sequence):
    root = mount / ('case-' + str(sequence))
    for suffix in ('', 'ro', 'rw', 'rw/nested', 'rw/nested/.git'):
        item = root / suffix
        item.mkdir(exist_ok=True)
        os.chown(item, 60000, 60000)
        os.chmod(item, 0o755)
    for suffix in ('ro/payload', 'rw/nested/.git/config', 'rw/secret', 'rw/plain'):
        owned_file(root / suffix)
    outside = base / ('host-source-' + str(sequence))
    owned_file(outside)
    os.symlink(root / 'rw/nested/.git/config', root / 'rw/protected-alias')
    os.symlink(root / 'rw/secret', root / 'rw/secret-alias')
    os.symlink(outside, root / 'rw/source-alias')
    shutil.copyfile(worker, root / 'rw/setuid-worker')
    os.chown(root / 'rw/setuid-worker', 0, 0)
    os.chmod(root / 'rw/setuid-worker', 0o4755)
    os.mknod(root / 'rw/null-device', stat.S_IFCHR | 0o666, os.stat('/dev/null').st_rdev)
    os.chmod(root / 'rw/null-device', 0o666)
    return root, outside


def profile_for(root):
    # Only generated absolute fixture paths, with Scheme string escaping.
    q = lambda path: json.dumps(str(path))
    ancestors = {path for item in (root, worker, pathlib.Path('/usr/lib'), pathlib.Path('/System/Library'), pathlib.Path('/dev/null')) for path in item.parents}
    metadata = ' '.join('(literal ' + q(path) + ')' for path in sorted(ancestors))
    return '\n'.join([
        '(version 1)', '(allow default)', '(deny network*)',
        '(deny mach-lookup)', '(deny mach-register)',
        '(deny file-read*)', '(deny file-write*)',
        '(allow file-read-metadata ' + metadata + ')',
        '(allow file-read* (subpath "/System/Library") (subpath "/usr/lib") (literal "/dev/null") (literal ' + q(worker) + ') (subpath ' + q(root) + '))',
        '(allow file-write* (subpath ' + q(root / 'rw') + '))',
        '(deny file-write* (subpath ' + q(root / 'rw/nested/.git') + '))',
        '(deny file-write-unlink (literal ' + q(root / 'rw/nested') + '))',
        '(deny file-read* file-write* (literal ' + q(root / 'rw/secret') + ') (literal ' + q(root / 'rw/absent-secret') + '))',
    ]) + '\n'


command(['/usr/bin/sw_vers'])
command(['/usr/bin/uname', '-a'])
command([worker, 'empty'])
command(['/usr/bin/hdiutil', 'create', '-size', '128m', '-fs', 'APFS', '-volname', 'CrucibleFSFixture', image])
_, data = command(['/usr/bin/hdiutil', 'attach', image, '-nobrowse', '-owners', 'on', '-mountpoint', mount, '-plist'])
entities = plistlib.loads(data)['system-entities']
mounted = [entry for entry in entities if entry.get('mount-point') == str(mount)]
assert len(mounted) == 1
_, data = command(['/usr/sbin/diskutil', 'info', '-plist', mount])
info = plistlib.loads(data)
assert info.get('FilesystemType') == 'apfs' and info.get('MountPoint') == str(mount)
assert info.get('DeviceNode') == mounted[0].get('dev-entry')
disks = [entry['dev-entry'] for entry in entities if entry.get('content-hint') == 'GUID_partition_scheme']
assert len(disks) == 1 and disks[0].startswith('/dev/disk')
command(['/sbin/mount', '-u', '-o', 'nosuid,nodev', mount])
command([worker, 'flags', mount])
operations = ['allowed-read', 'allowed-write', 'source-read', 'source-write', 'readonly-write',
              'protected-write', 'protected-unlink', 'protected-rename', 'ancestor-rename',
              'protected-link', 'protected-symlink', 'source-symlink', 'unreadable-read',
              'unreadable-alias', 'unreadable-case', 'unreadable-create', 'unreadable-rename', 'device', 'setuid']
failures = []
sequence = 0
for mode in ('control', 'confined'):
    for operation in operations:
        command([worker, 'empty'])
        root, outside = make_case(sequence)
        policy = base / ('profile-' + str(sequence) + '.sb')
        policy.write_text(profile_for(root))
        os.chmod(policy, 0o644)
        code, data = command([worker, 'launch', mode, policy, operation, root, outside], allowed=(0, 1, 77))
        # Every guest fixture is single-process, including exec of the setuid worker.
        # The parent wait reaps it. A new case requires an error-checked empty UID.
        command([worker, 'empty'])
        if code:
            failures.append({'mode': mode, 'operation': operation, 'status': code})
        print('RESULT ' + json.dumps({'mode': mode, 'operation': operation, 'status': code}), flush=True)
        sequence += 1
command([worker, 'unmount', mount])
command(['/usr/bin/hdiutil', 'detach', disks[0]])
print('COMPLETE cases=' + str(sequence) + ' failures=' + json.dumps(failures) + ' uid_empty=1 nonforced_detach=1', flush=True)
assert not failures, failures
# Retain exact owned fixture until disposable VM destruction; never generic cleanup.

File: .github/probes/macos-filesystem.c
#define _DARWIN_C_SOURCE 1
#include <dlfcn.h>
#include <errno.h>
#include <fcntl.h>
#include <grp.h>
#include <libproc.h>
#include <limits.h>
#include <pwd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mount.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <sys/syscall.h>
#include <sys/types.h>
#include <unistd.h>

static const uid_t guest_uid=60000;
static void fail(const char *why) { perror(why); exit(77); }
static void path(char *out, const char *root, const char *suffix) {
    int n=snprintf(out,PATH_MAX,"%s/%s",root,suffix);
    if(n<0 || n>=PATH_MAX) { errno=ENAMETOOLONG; fail("path"); }
}
static void identity(void) {
    struct proc_bsdinfo ids;
    if(proc_pidinfo(getpid(),PROC_PIDTBSDINFO,0,&ids,sizeof(ids))!=(int)sizeof(ids)) fail("identity query");
    if(ids.pbi_ruid!=guest_uid || ids.pbi_uid!=guest_uid || ids.pbi_svuid!=guest_uid ||
       ids.pbi_rgid!=guest_uid || ids.pbi_gid!=guest_uid || ids.pbi_svgid!=guest_uid) {
        fprintf(stderr,"IDENTITY-MISMATCH real=%u effective=%u saved=%u\n",ids.pbi_ruid,ids.pbi_uid,ids.pbi_svuid); exit(77);
    }
    gid_t groups[64];
    int (*kernel_call)(int, ...)=dlsym(RTLD_DEFAULT,"syscall");
    if(!kernel_call) fail("kernel group API");
    int count=kernel_call(SYS_getgroups,64,groups);
    if(count!=1 || groups[0]!=guest_uid) fail("kernel groups");
    errno=0;
    if(setuid(0)!=-1 || errno!=EPERM) fail("root regain");
}
static void empty_uid(void) {
    if(getpwuid(guest_uid)!=NULL) { errno=EEXIST; fail("fixture UID account exists"); }
    pid_t pids[4096]; errno=0;
    int n=proc_listpids(PROC_RUID_ONLY,guest_uid,pids,sizeof(pids));
    if(n<0 || n>=(int)sizeof(pids) || n%(int)sizeof(pid_t) || (n==0 && errno)) fail("UID enumeration");
    if(n!=0) { fprintf(stderr,"UID not empty bytes=%d\n",n); exit(77); }
    puts("UID-EMPTY");
}
static void launch(int argc, char **argv) {
    if(argc!=7 || geteuid()!=0) exit(77);
    gid_t group=guest_uid;
    struct rlimit files={128,128}, processes={8,8}, size={1024*1024,1024*1024}, core={0,0};
    if(setrlimit(RLIMIT_NOFILE,&files) || setrlimit(RLIMIT_NPROC,&processes) ||
       setrlimit(RLIMIT_FSIZE,&size) || setrlimit(RLIMIT_CORE,&core) ||
       setgroups(1,&group) || setgid(guest_uid) || setuid(guest_uid)) fail("permanent drop");
    identity(); alarm(10);
    struct proc_fdinfo descriptors[128];
    int bytes=proc_pidinfo(getpid(),PROC_PIDLISTFDS,0,descriptors,sizeof(descriptors));
    if(bytes<=0 || bytes>=(int)sizeof(descriptors) || bytes%(int)sizeof(descriptors[0])) fail("descriptor inventory");
    for(int i=0;i<bytes/(int)sizeof(descriptors[0]);i++) if(descriptors[i].proc_fd>2) { errno=EBADF; fail("unexpected inherited descriptor"); }
    if(strcmp(argv[2],"confined")==0) {
        FILE *f=fopen(argv[3],"rb"); if(!f) fail("profile read");
        char profile[16385]; size_t n=fread(profile,1,sizeof(profile)-1,f);
        if(ferror(f) || !feof(f) || !n) fail("profile bounds");
        profile[n]=0; fclose(f);
        int (*initialize)(const char *, unsigned long, char **)=dlsym(RTLD_DEFAULT,"sandbox_init");
        void (*free_error)(char *)=dlsym(RTLD_DEFAULT,"sandbox_free_error");
        if(!initialize || !free_error) fail("sandbox API unavailable");
        char *error=NULL;
        puts("LAUNCH before-profile");
        if(initialize(profile,0,&error)) { fprintf(stderr,"PROFILE %.256s\n",error?error:"unknown"); if(error)free_error(error); exit(77); }
        puts("LAUNCH after-profile");
    } else if(strcmp(argv[2],"control")!=0) exit(77);
    /* No untrusted code runs in this pre-policy image. Fresh exec is mandatory. */
    char *args[]={argv[0],"guest",argv[2],argv[4],argv[5],argv[6],NULL};
    char *environment[]={"PATH=/usr/bin:/bin","LANG=C",NULL};
    puts("LAUNCH before-exec");
    execve(argv[0],args,environment); fail("guest exec");
}
static int operation(const char *name, const char *root, const char *outside) {
    char a[PATH_MAX],b[PATH_MAX];
    if(!strcmp(name,"allowed-read")) { path(a,root,"ro/payload"); return open(a,O_RDONLY); }
    if(!strcmp(name,"allowed-write")) { path(a,root,"rw/new"); return open(a,O_CREAT|O_EXCL|O_WRONLY,0600); }
    if(!strcmp(name,"source-read")) return open(outside,O_RDONLY);
    if(!strcmp(name,"source-write")) return open(outside,O_WRONLY);
    if(!strcmp(name,"readonly-write")) { path(a,root,"ro/payload"); return open(a,O_WRONLY); }
    if(!strcmp(name,"protected-write")) { path(a,root,"rw/nested/.git/config"); return open(a,O_WRONLY); }
    if(!strcmp(name,"protected-unlink")) { path(a,root,"rw/nested/.git/config"); return unlink(a); }
    if(!strcmp(name,"protected-rename")) { path(a,root,"rw/nested/.git"); path(b,root,"rw/moved-git"); return rename(a,b); }
    if(!strcmp(name,"ancestor-rename")) { path(a,root,"rw/nested"); path(b,root,"rw/moved-parent"); return rename(a,b); }
    if(!strcmp(name,"protected-link")) { path(a,root,"rw/nested/.git/config"); path(b,root,"rw/linked-config"); return link(a,b); }
    if(!strcmp(name,"protected-symlink")) { path(a,root,"rw/protected-alias"); return open(a,O_WRONLY); }
    if(!strcmp(name,"source-symlink")) { path(a,root,"rw/source-alias"); return open(a,O_RDONLY); }
    if(!strcmp(name,"unreadable-read")) { path(a,root,"rw/secret"); return open(a,O_RDONLY); }
    if(!strcmp(name,"unreadable-alias")) { path(a,root,"rw/secret-alias"); return open(a,O_RDONLY); }
    if(!strcmp(name,"unreadable-case")) { path(a,root,"rw/SECRET"); return open(a,O_RDONLY); }
    if(!strcmp(name,"unreadable-create")) { path(a,root,"rw/absent-secret"); return open(a,O_CREAT|O_EXCL|O_WRONLY,0600); }
    if(!strcmp(name,"unreadable-rename")) { path(a,root,"rw/plain"); path(b,root,"rw/absent-secret"); return rename(a,b); }
    if(!strcmp(name,"device")) { path(a,root,"rw/null-device"); return open(a,O_RDWR); }
    if(!strcmp(name,"setuid")) {
        path(a,root,"rw/setuid-worker"); execl(a,a,"identity",NULL); return -1;
    }
    errno=EINVAL; return -1;
}
int main(int argc,char **argv) {
    setvbuf(stdout,NULL,_IONBF,0);
    if(argc==2 && !strcmp(argv[1],"empty")) { empty_uid(); return 0; }
    if(argc==2 && !strcmp(argv[1],"identity")) { alarm(10); identity(); puts("SETUID-REFUSED"); return 0; }
    if(argc==3 && !strcmp(argv[1],"flags")) {
        struct statfs fs; if(statfs(argv[2],&fs)) fail("statfs");
        printf("FLAGS nosuid=%d nodev=%d readonly=%d\n",!!(fs.f_flags&MNT_NOSUID),!!(fs.f_flags&MNT_NODEV),!!(fs.f_flags&MNT_RDONLY));
        return (fs.f_flags&(MNT_NOSUID|MNT_NODEV))==(MNT_NOSUID|MNT_NODEV)?0:77;
    }
    if(argc==3 && !strcmp(argv[1],"unmount")) { if(unmount(argv[2],0)) fail("nonforced unmount"); return 0; }
    if(argc>1 && !strcmp(argv[1],"launch")) launch(argc,argv);
    if(argc!=6 || strcmp(argv[1],"guest")) return 77;
    puts("GUEST entered");
    alarm(10); identity();
    errno=0; int result=operation(argv[3],argv[4],argv[5]); int error=errno;
    int allow=!strcmp(argv[2],"control") || !strcmp(argv[3],"allowed-read") || !strcmp(argv[3],"allowed-write");
    if(!strcmp(argv[3],"device")) allow=0;
    /* setuid success replaces this process; only refusal to exec reaches here. */
    printf("CASE mode=%s name=%s result=%d errno=%d expected=%s\n",argv[2],argv[3],result,error,allow?"allow":"deny");
    if(allow) return result>=0?0:1;
    return result==-1 && (error==EACCES || error==EPERM || (!strcmp(argv[3],"device") && error==ENXIO))?0:1;
}
